bftengine/src/bftengine/Crypto.cpp

```diff
@@ -58,24 +58,27 @@ namespace impl {
 //#define RSA_STANDARD PKCS1v15
 //#define RSA_STANDARD OAEP<SHA>
 
-static RandomPool sGlobalRandGen;  // not thread-safe !!
+// Wrap RandomPool in a struct which initialises it. The wrapper is used as static thread local variable so that each
+// thread has got its own instance. The reason for this complication is thread safety. This RandomPool is used by the
+// SigManager which is a singleton used by many threads concurrently.
+struct RandomPoolWrapper : RandomPool {
+  RandomPoolWrapper() { CryptographyWrapper::init(); }
+};
+
+static thread_local RandomPoolWrapper sGlobalRandGen;
 
-void CryptographyWrapper::init(const char* randomSeed) {
-  string s(randomSeed);
-  if (s.length() < 16) s.resize(16, ' ');
-  sGlobalRandGen.IncorporateEntropy((CryptoPP::byte*)s.c_str(), s.length());
+void CryptographyWrapper::init() {
+  SecByteBlock seed(32);
+  // this might block for a while - man 4 random for details
+  OS_GenerateRandomBlock(false, seed, seed.size());
+  sGlobalRandGen.IncorporateEntropy(seed, seed.size());
 
   VERIFY(DigestUtil::digestLength() == DIGEST_SIZE);
 
   // Initialize RELIC library for BLS threshold signatures
   BLS::Relic::Library::Get();
```

Comment:
I wonder if we need this here... If not the whole init() can be moved inside RandomPoolWrapper constructor.